modules/database/src/ioc/misc/iocInit.c

```diff
@@ -546,28 +557,34 @@ static void doResolveLinks(dbRecordType *pdbRecordType, dbCommon *precord,
 
         dbInitLink(plink, pdbFldDes->field_type);
     }
+    return 0;
 }
 
-static void doInitRecord1(dbRecordType *pdbRecordType, dbCommon *precord,
+static long doInitRecord1(dbRecordType *pdbRecordType, dbCommon *precord,
     void *user)
 {
     rset *prset = pdbRecordType->prset;
 
-    if (!prset) return;         /* unlikely */
+    if (!prset) return S_dbLib_noRecSup;         /* unlikely */
 
     if (prset->init_record)
         prset->init_record(precord, 1);
+    return 0;
 }
 
-static void initDatabase(void)
+static long initDatabase(void)
 {
+    long ret;
     dbChannelInit();
-    iterateRecords(doInitRecord0, NULL);
-    iterateRecords(doResolveLinks, NULL);
-    iterateRecords(doInitRecord1, NULL);
-
-    epicsAtExit(exitDatabase, NULL);
-    return;
+    ret = iterateRecords(doInitRecord0, NULL);
+    if(!ret)
+        ret = iterateRecords(doResolveLinks, NULL);
+    if(!ret)
+        ret = iterateRecords(doInitRecord1, NULL);
+
+    if(!ret)
+        ret = epicsAtExit(exitDatabase, NULL);
+    return ret;
```

Comment:
> The major behavior change with this commit is ...

Avoiding potential for change is why I explicitly didn't add a test for eg. `init_record()` returning an error.  So I believe the only previously "soft" errors which would now become "hard" errors have to do with link initialization and missing device/driver support tables (imo. a serious error).

> ... that's called graceful degradation ...

Also a trap for the unwary...  Our default is like [`On Error Resume Next`](https://docs.microsoft.com/en-us/dotnet/visual-basic/language-reference/statements/on-error-statement).

Reply:
From what I see, the only non-success return values come from the new EXT addition and from missing `prset`s, so I don't expect breakage of databases that are not already broken.

Reply:
I just noticed this again while re-reviewing the code. Should we add something like a "failFast" flag, which would tell the IOC whether to stop initialization on the first error it finds or to continue anyway? Operational IOCs might have that cleared, but that statement would have to be explicitly added to the st.cmd file, and by default it would be set. Unfortunately that would require more tests to check that things still run okay with it cleared.

Reply:
To elaborate on my previous comment: I have looked at `dbInitLink()`, which IIUC is the function you are worried about, and it returns success in all cases except in the case of a missing EXT modifier added by this PR. So while I think a "failFast" flag would be useful if new checks are added, it is not needed for this PR because it is an opt-in change already.